stdlib/universe/fill.go

```diff
@@ -204,59 +214,63 @@ func (t *fillTransformation) Process(id execute.DatasetID, tbl flux.Table) error
 		}
 	}
 
-	builder, created := t.cache.TableBuilder(key)
-	if created {
-		if err := execute.AddTableCols(tbl, builder); err != nil {
-			return err
-		}
-	}
-	idx := execute.ColIdx(t.spec.Column, builder.Cols())
-	if idx < 0 {
-		return errors.Newf(codes.FailedPrecondition, "fill column not found: %s", t.spec.Column)
-	}
-
 	prevNonNull := t.spec.Value
 	if !t.spec.UsePrevious {
-		if builder.Cols()[idx].Type != flux.ColumnType(prevNonNull.Type()) {
-			return errors.Newf(codes.FailedPrecondition, "fill column type mismatch: %s/%s", builder.Cols()[idx].Type.String(), flux.ColumnType(prevNonNull.Type()).String())
+		if tbl.Cols()[idx].Type != flux.ColumnType(prevNonNull.Type()) {
+			return errors.Newf(codes.FailedPrecondition, "fill column type mismatch: %s/%s", tbl.Cols()[idx].Type.String(), flux.ColumnType(prevNonNull.Type()).String())
 		}
 	}
-	return tbl.Do(func(cr flux.ColReader) error {
-		for j := range cr.Cols() {
-			if j == idx {
-				continue
-			}
-			if err := execute.AppendCol(j, j, cr, builder); err != nil {
-				return err
-			}
-		}
-		// Set new value
-		l := cr.Len()
 
-		if l > 0 {
-			if t.spec.UsePrevious {
-				prevNonNull = execute.ValueForRow(cr, 0, idx)
+	table, err := table.StreamWithContext(t.ctx, key, tbl.Cols(), func(ctx context.Context, w *table.StreamWriter) error {
+		return tbl.Do(func(cr flux.ColReader) error {
+			l := cr.Len()
+			if l == 0 {
+				return nil
 			}
-
-			for i := 0; i < l; i++ {
-				v := execute.ValueForRow(cr, i, idx)
-				if v.IsNull() {
-					if err := builder.AppendValue(idx, prevNonNull); err != nil {
-						return err
-					}
+			vs := make([]array.Interface, len(w.Cols()))
+			for j, col := range w.Cols() {
+				if j != idx {
+					vs[j] = table.Values(cr, j)
+					vs[j].Retain()
 				} else {
-					if err := builder.AppendValue(idx, v); err != nil {
-						return err
-					}
 					if t.spec.UsePrevious {
-						prevNonNull = v
+						prevNonNull = execute.ValueForRow(cr, 0, idx)
+					}
+					b := arrow.NewBuilder(col.Type, t.alloc)
+					b.Reserve(l)
+					hasNull := false
+					for i := 0; i < l; i++ {
+						v := execute.ValueForRow(cr, i, idx)
```

Comment:
You should avoid `execute.ValueForRow`. It is slow because it calls into an interface, performs a checked type cast, then it will allocate memory to place the retrieve value into an interface to return it which always requires an allocation. Since you are processing a single column, you can write a separate implementation for each column type.

The `filter()` transformation has a good example (https://github.com/influxdata/flux/blob/master/internal/arrowutil/filter.gen.go.tmpl) and so does `pivot()` (https://github.com/influxdata/flux/blob/master/stdlib/universe/pivot.gen.go.tmpl).

For this circumstance specifically, `pivot()` is the better example to mimic.